### PR TITLE
Btrfs optimizations for storing and deleting layers

### LINKED ISSUE #35630: btrfs graphdriver always create writable layers
**Description**

According to comments in various implementations of graphdriver meant that _CreateReadWrite_ function is create writable layer and _Create_ read-only. But in virtually all implementation this functions do exactly the same.
For btrfs graphdriver is important to have as much as possible read-only layers because they are available in file tree. Any misc system action which traverses the file tree will cause of updating access time of all files in all writable subvolumes. For complex images it cost a lot of IO and explosive growth of metadata size.
Also there is a risk of easy damage of intermediate layer. If it's happen between pull of two images with shared layers integrity of second one will be silently broken.

**Steps to reproduce the issue:**
1. Clean system with no images
```
# btrfs f df /var/lib/docker/
Data, single: total=2.00GiB, used=648.00KiB
System, single: total=32.00MiB, used=16.00KiB
Metadata, single: total=256.00MiB, used=224.00KiB
GlobalReserve, single: total=16.00MiB, used=0.00B
```
2. Pull any complex image
```
# ls -1 /var/lib/docker/btrfs/subvolumes/ | wc -l
37
# btrfs f df /var/lib/docker/
Data, single: total=2.00GiB, used=1.42GiB
System, single: total=32.00MiB, used=16.00KiB
Metadata, single: total=256.00MiB, used=95.69MiB
GlobalReserve, single: total=32.00MiB, used=0.00B
```
Metadata in expected borders
3. Now init any system process which will walks file tree, selinux retag or something like this.
```
# sosreport -a --batch
# btrfs f df /var/lib/docker/
Data, single: total=2.00GiB, used=1.42GiB
System, single: total=32.00MiB, used=16.00KiB
Metadata, single: total=1.50GiB, used=1.32GiB
GlobalReserve, single: total=64.00MiB, used=0.00B
```

**Describe the results you received:**
Operation take a  significant longer time and cause unexpected write IO.
btrfs metadata size growth dramaticaly.
All /var/lib/docker/btrfs/subvolumes/* writable

**Describe the results you expected:**
No extra write. No btrfs metadata changes, we just read files.
Intermediate layers in /var/lib/docker/btrfs/subvolumes/ are read only, only container layers writable.

**Output of `docker version`:**

```
Client:
 Version:      17.06.2-ol
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   d02b7ab
 Built:        Fri Oct  6 00:02:23 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.2-ol
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   d02b7ab
 Built:        Fri Oct  6 00:03:48 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 17.06.2-ol
Storage Driver: btrfs
 Build Version: Btrfs v4.9.1
 Library Version: 102
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: 810190ceaa507aa2727d7ae6f4790c76ec150bd2
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.1.12-103.9.7.el7uek.x86_64
Operating System: Oracle Linux Server 7.4
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.943GiB
Name: op-vm4
ID: BBAC:TM7H:ZTFL:YXE5:5AAT:PNVC:GEGI:2774:LYFA:4PN6:6DMB:OMLW
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 16
 Goroutines: 25
 System Time: 2017-11-28T20:15:53.627776456+03:00
 EventsListeners: 0
Http Proxy: http://192.168.100.52:3128
Https Proxy: http://192.168.100.52:3128
No Proxy: localhost,127.0.0.1,192.168.100.27
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 192.168.100.27:5000
 127.0.0.0/8
Live Restore Enabled: false
```

### PR DESCRIPTION
**Pull request by saiarcot895**
**- What I did**

This changeset is a cherry-pick of the changes in #36536. Specifically, the changes related to optimizing deleting layers are included here. Changes related to quotas are not included here, as I don't use quotas and so can't reliably test the changes.

Changes related to storing layers as read-only in btrfs are not included, due to the behavior change that would be required if deleting the Docker data directory entirely on btrfs filesystems.

**- How I did it**

**- How to verify it**

I've tested the resulting dockerd binary and after deleting all caches and layers, I was able to run `make binary` for Docker. Additionally, a quick smoke test with some public containers seems to work fine. I've also verified with `strace` that when removing containers and layers, Docker is no longer traversing the entire layer.

**- Description for the changelog**

In summary, the changes in this branch will do the following:
1. Change the subvolume deletion code to ask btrfs for any nested subvolumes using ioctl calls, instead of traversing each layer. This will result in a huge performance improvement for large layers, as the number of system calls will be much lower.

With these changes, rootless mode appears to still work (at least for the basic container pull, start, and delete cases). However, container deletion in rootless mode will be significantly slower than in root mode, because the ioctl call to delete a subvolume (and check for nested subvolumes) requires root permissions. This, I believe, is the same as the current code.

Fixes #42973

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Do you have benchmark data?

**Comment by saiarcot895** [comment]
For the `mysql:latest`, `python:3.9-bullseye`, and `golang:bullseye` images, deleting all three images (as part of a single command) takes 6.1 seconds without these changes, and 3.4 seconds with these changes.

For a 5.4GB image with 62 layers (used as a build environment), deleting that images takes 2 minutes 45 seconds without these changes, and 8.4 seconds with my changes

**Comment by cpuguy83** [comment]
Seems like the graphdriver interface should be updated to add a `Commit` function that signals that the layer should be marked as read-only... and probably `CreateReadWrite` removed since every layer needs to be created as read-write to begin with anyway.

Adding special cleanup code for btrfs in the tests seems ok.

I do wonder how the proposed change affects users expecting to be able to `rm -rf /var/lib/docker`.

**Comment by saiarcot895** [comment]
> I do wonder how the proposed change affects users expecting to be able to `rm -rf /var/lib/docker`.

Yeah, `rm -rf /var/lib/docker` would no longer work (even with root permissions) without first marking the subvolumes as not read-only. Would that be too much of a breaking change?

**Comment by saiarcot895** [comment]
> > I do wonder how the proposed change affects users expecting to be able to `rm -rf /var/lib/docker`.
> 
> Yeah, `rm -rf /var/lib/docker` would no longer work (even with root permissions) without first marking the subvolumes as not read-only. Would that be too much of a breaking change?

Would this be considered a breaking change and not allowed? If so, should I remove this part of the change?

**Comment by saiarcot895** [comment]
@cpuguy83 @thaJeztah PTAL

**Comment by saiarcot895** [comment]
Closing this PR due to age and opening a new one.
